Services/Devices.cs

```diff
@@ -71,14 +72,16 @@ public class Devices : IDevices
             this.connectionStringManager = connStringManager;
             this.twinReadsWritesEnabled = config.TwinReadWriteEnabled;
             this.registryCount = -1;
-            this.SetCurrentIotHub();
+            this.setupDone = false;
         }
-        
+
         /// <summary>
         /// Ping the registry to see if the connection is healthy
         /// </summary>
         public async Task<Tuple<bool, string>> PingRegistryAsync()
         {
+            this.SetupHub();
```

Comment:
Why move this outside the constructor?  It looks like every function needs it so that seems like the right spot.

Reply:
the function can throw an exception, which is fine when the exception occurs in a method. If the exception occurs in the ctor instead, it breaks DI, for instance the StatusController would be unable to report that IoTHub is not working. I added a comment in the code to clarify this. Generally speaking, we should avoid dealing with external dependencies in the constructors, for the same reason, we can use lazy initialization or this kind of construct.